### PR TITLE
Redesign IVR menu with updated pagination and key functionality

### PR DESCRIPTION
**Pull request by farmanahmed888**
This pull request updates the IVR menu navigation key system, removing the "repeat current menu" option and changing how pagination works in menu navigation. The changes affect both the core FSM instantiation logic and the associated constants, and they are thoroughly tested with new and updated unit tests.

Key changes include:

**Navigation Key Redesign and Removal of Repeat Option:**

- The `repeat_current_categories_key` and its associated menu/transition logic have been removed from both `insti.py` (menu navigation) and `pure_audio.py` (audio playback state). This means users can no longer press a key to repeat the current menu or audio menu. [[1]](diffhunk://#diff-a7fc0478b7b950b67a72dba3d9f35bfb8c94b49d4194c8cc8ec38d25f9af8801L33-L34) [[2]](diffhunk://#diff-a7fc0478b7b950b67a72dba3d9f35bfb8c94b49d4194c8cc8ec38d25f9af8801L280-L284) [[3]](diffhunk://#diff-a7fc0478b7b950b67a72dba3d9f35bfb8c94b49d4194c8cc8ec38d25f9af8801L432-L439) [[4]](diffhunk://#diff-d515420b85011e353509ddd7aadc65027488cb83f2d7a24f533bcb028dbb3b09L20) [[5]](diffhunk://#diff-d515420b85011e353509ddd7aadc65027488cb83f2d7a24f533bcb028dbb3b09L129-R129) [[6]](diffhunk://#diff-d515420b85011e353509ddd7aadc65027488cb83f2d7a24f533bcb028dbb3b09L174-L181)
- The constants for navigation keys have been updated: the menu page size is now 8, the "next page" key is now `#`, and the "previous page" key is `*`. The repeat key constant is fully removed.

**Refactoring and Type Improvements:**

- The `_Option` class now accepts both `int` and `str` types for its `key`, improving flexibility for navigation keys.
- The mapping logic for navigation keys in `_add_nav_action` now handles both numeric and string keys robustly.

**Testing and Validation:**

- Comprehensive unit tests have been added for the new menu navigation logic (`test_insti_nav_keys.py`), for the updated IVR constants (`test_ivr_constants.py`), and for the removal of the repeat key in audio playback (`test_pure_audio_nav_keys.py`). These ensure correct behavior for pagination, key mapping, and transition logic after the redesign. [[1]](diffhunk://#diff-62edc2d7867bd6e8cbc4271ee4272e5d40550ce7b6429f892ee96f9d8a8927e0R1-R74) [[2]](diffhunk://#diff-165363cccc6501d9a5cc101ff2f5c438a607fb39caaa74f28b839d07e85d9265R1-R35) [[3]](diffhunk://#diff-9cd0176f3a5094485911ffaaf611ce1faf5f26067c6e066e5bfb66b49a3eabedR1-R75)

**Constants Cleanup:**

- The `repeatCurrentMenuUrl` and `repeat_current_categories_key` constants are removed from `ivr_constants.py`, as they are no longer needed. [[1]](diffhunk://#diff-2e6e106e299ee7bd28d5e7a5e712507b2357771fa016c166941dadf6134d802dL102) [[2]](diffhunk://#diff-2e6e106e299ee7bd28d5e7a5e712507b2357771fa016c166941dadf6134d802dL111-R112)

These changes modernize the menu navigation experience, simplify the codebase, and ensure correctness through targeted tests.

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
### Key Findings

- The pagination redesign introduces a **likely correctness regression** in menu navigation by changing the page size and navigation keys without verifying downstream consumers of the option map.
- The `_Option` type relaxation is fine in principle, but the new mixed-key mapping can create **hard-to-debug collisions** if numeric and string keys converge on the same logical value.
- The removed repeat action appears intentionally covered by tests, but the playback state change should be validated against any callers expecting the old `repeat` option semantics.

### Issues

1. **HIGH — `platform/app/services/fsm/instantiation/insti.py`, function `_add_nav_action`**
   - The new mixed `int | str` key handling can still produce ambiguous mappings when a string key is numeric-like and an integer key with the same value exists elsewhere in the same menu generation path. That risks overwriting entries in `key_to_value_mapping` and routing the wrong label/action.
   - **Concrete fix:** Normalize all navigation keys to a single canonical type before insertion, and enforce uniqueness when building the mapping. If string DTMF keys are required, keep both the menu option key and mapping key as strings throughout.

2. **HIGH — `platform/app/services/fsm/instantiation/pure_audio.py`, function `generate_state`**
   - The option label for key `0` was changed to a pause/resume concept, but no transition logic in this diff shows that the corresponding behavior was added. If the state machine still routes `0` to the previous “instructions to exit” flow or to no-op, users will get a misleading menu entry.
   - **Concrete fix:** Add or update the `0` transition and handler so the action matches the new label, or revert the label until the behavior is implemented.

3. **MEDIUM — `platform/app/services/fsm/instantiation/insti.py`, function `_get_stream_actions`**
   - Increasing the page size from 4 to 8 changes the navigation surface without any validation that all menu consumers, audio prompts, and keypad UX still fit the new range. This is especially risky because the tests only cover local option presence, not end-to-end state transitions.
   - **Concrete fix:** Add an integration-level test that exercises a real multi-page menu and verifies `#`/`*` transitions, selected content routing, and audio prompt generation across page boundaries.

4. **LOW — `platform/app/services/fsm/instantiation/ivr_constants.py`, module-level constants**
   - The navigation constants are now hard-coded in multiple tests as well as the module. That increases drift risk if the page size or DTMF keys change again.
   - **Concrete fix:** Centralize any repeated assertions in tests by importing the constants directly and avoid duplicating semantic expectations in helper names or inline comments.

### Review Summary

| Severity | Open Issues |
|----------|-------------|
| Critical | 0 |
| High | 2 |
| Medium | 1 |
| Low | 1 |

**Score:** 89/100

**Review by a4i-architect — COMMENTED** [review]
### Key findings

- The navigation redesign is mostly consistent, but it introduces a **likely broken pager flow** for any menu with more than 8 items.
- The new key handling in `insti.py` and `pure_audio.py` is directionally correct, but there are **gaps in how navigation transitions are modeled** after removing repeat.
- The new tests cover the intended behavior well, but they also appear to **codify assumptions that may not hold across larger menus**.

### Issues

#### 1) `platform/app/services/fsm/instantiation/insti.py` — `_get_stream_actions`
**High:** Pagination keys are now added only as menu options, but the transition generation for `#` / `*` is not visible in the modified logic. If those keys are not mapped to state transitions elsewhere, the second page cannot actually be reached or navigated back from, which breaks menus larger than one page.  
**Concrete fix:** Add explicit FSM transitions for the new pagination keys in the same place where `previous_category_level_key` and category transitions are registered, and ensure the destination state IDs match the page index logic.

#### 2) `platform/app/services/fsm/instantiation/insti.py` — `_add_nav_action`
**Medium:** The new `mapping_key` logic changes the key type from always-`int` to sometimes-`str`, but this function still depends on `str(nav_key).isdigit()`, which silently treats any non-numeric key as a raw string. That is fine for `#` and `*`, but it makes the mapping behavior implicit and easy to misuse for future keys, especially if single-character numeric-like strings are introduced.  
**Concrete fix:** Replace the ad hoc type coercion with an explicit key-normalization helper that documents allowed navigation key formats and raises on unsupported keys.

#### 3) `platform/app/services/fsm/instantiation/pure_audio.py` — `generate_state`
**High:** Removing the repeat transition also removed the only explicit self-loop on the playback state. If the intended replacement is “pause/resume” on `0`, there is no corresponding transition added here, so the menu option and FSM behavior can diverge.  
**Concrete fix:** Add a transition for the new `0` behavior, or remove the option until the FSM action is implemented, so the menu does not expose an inert key.

#### 4) `platform/app/services/fsm/instantiation/ivr_constants.py` — module constants
**Medium:** Changing `number_of_categories_listed_in_one_state` from 4 to 8 affects every consumer of the constant, but this PR only updates the menu generation path. Any other code that assumes the old page size will now behave inconsistently.  
**Concrete fix:** Audit all callers of `number_of_categories_listed_in_one_state` and update any paging, prompt text, or tests that still assume 4-item pages.

#### 5) `platform/tests/unit/test_pure_audio_nav_keys.py` — `_build_fsm_with_playback_state`
**Low:** The test setup hard-codes a very specific FSM shape and patches internal settings in a way that is tightly coupled to current implementation details. This makes the test brittle if the playback state ID or parent wiring changes.  
**Concrete fix:** Build the test FSM through a small helper that derives expected IDs from the same public state construction inputs instead of asserting on hard-coded IDs.

### Review Summary

| Severity | Open Issues |
|----------|-------------|
| Critical | 0 |
| High | 2 |
| Medium | 2 |
| Low | 1 |

**Score:** 87/100

**Review by a4i-architect — COMMENTED** [review]
### Key Findings

- The pagination redesign in `insti.py` and `ivr_constants.py` is broadly consistent, but it introduces a likely navigation regression around page sizing and keypad reuse.
- `pure_audio.py` now repurposes key `0` for pause/resume while removing repeat; that changes user-facing behavior and should be verified against the FSM transitions to avoid dead ends.
- The new tests cover the happy path well, but they do not validate that the removed repeat transition no longer leaves stale references in any downstream state generation.

### Issues

#### 1) `platform/app/services/fsm/instantiation/insti.py` — `_get_stream_actions`
**High**: The page size was increased to 8, but the navigation model still uses digit keys for both content selection and page controls. That makes the last content item and the paging controls compete for the same keypad space in edge cases, especially when a page is full and `#`/`*` are introduced later in the same menu lifecycle.

**Concrete fix:** Reserve page-control keys before assigning content keys, and compute the displayed content window from that reserved set so a full page never gets reinterpreted when pagination keys are appended.

#### 2) `platform/app/services/fsm/instantiation/pure_audio.py` — `generate_state`
**High**: Repurposing key `0` from “next/instructions to exit” to “pause/resume” changes the call flow but the transition table only shows exit on `9` and removal of repeat. If no transition is wired for `0`, callers will hear an option that cannot be acted on.

**Concrete fix:** Add an explicit FSM transition for `0` or remove the menu option until the state machine supports pause/resume end-to-end.

#### 3) `platform/app/services/fsm/instantiation/insti.py` — `_Option`
**Medium**: Changing `key` to `str` is correct for DTMF handling, but this class is still manually typed and can now accept invalid values without validation. That increases the risk of inconsistent menu dictionaries if a caller passes a non-string-like key.

**Concrete fix:** Make `_Option` a typed dataclass or validate `key` at construction time with a narrow `str` type and explicit conversion at the call sites.

#### 4) `platform/tests/unit/test_insti_nav_keys.py` — `test_nine_items_second_page_has_prev_only`
**Medium**: The new test assumes page 2 state indexing via `state=1` without asserting the FSM’s page offset contract. If the generator changes its pagination offset semantics, this test may pass or fail for the wrong reason.

**Concrete fix:** Assert the exact page window content returned for `state=1` and verify the generated transitions, not just the menu option labels.

#### 5) `platform/tests/unit/test_pure_audio_nav_keys.py` — `_build_fsm_with_playback_state`
**Low**: The helper relies on patched settings and magic state IDs that are tightly coupled to current implementation details. That makes the test brittle and harder to maintain as the FSM naming convention evolves.

**Concrete fix:** Build the helper through a small test fixture or factory that derives the expected state ID from the same public inputs used by `generate_state`.

### Review Summary

| Severity | Open Issues |
|----------|-------------|
| Critical | 0 |
| High | 2 |
| Medium | 2 |
| Low | 1 |

**Score:** 86/100

File: platform/app/services/fsm/instantiation/ivr_constants.py [hunk 1]
```diff
@@ -108,10 +107,9 @@ def get_content_url() -> str:
 audioFinishedMessageUrl = "audioDialogs/{language}/audioFinishedDialog/{speechRate}.mp3"
 
 # Navigation key constants
-number_of_categories_listed_in_one_state = 4
-next_n_categories_key = "5"
-previous_n_categories_key = "7"
-repeat_current_categories_key = "8"
+number_of_categories_listed_in_one_state = 8
+next_n_categories_key = "#"
```
Comment (theainerd) [inline platform/app/services/fsm/instantiation/ivr_constants.py:111]: **Blocking:** a raw `#` here corrupts the generated audio URL.

`_get_key_press_url` (`insti.py:83`) interpolates the key into the blob path, producing:

```
https://<acct>.blob.core.windows.net/pull-model-menus/pressKeysDialog/english/#/1.0.mp3
```

`#` is the URL fragment delimiter, so `SASService.get_url_with_sas` (`sas_service.py:98-107`) does `urlparse(unquote(url))` and gets `path='/pull-model-menus/pressKeysDialog/english/'` with `fragment='/1.0.mp3'`. It then signs the blob `pressKeysDialog/english` and returns `.../pressKeysDialog/english?sv=...`.

I confirmed this by executing `_get_stream_actions` on this branch — that exact URL is emitted.

Impact: any menu with more than 8 items (e.g. a language with 9+ themes) streams a BlobNotFound instead of the "press #" prompt, so the caller is never told how to reach page 2.

Fix: percent-encode as `%23` in the URL path (and make sure a blob exists at the corresponding name).

**Review by theainerd — COMMENTED** [review]
One blocking issue on the new navigation keys; details inline.

**Review by mamuqsit — APPROVED** [review]

**Review by a4i-architect — COMMENTED** [review]
### High-level summary
This PR mostly cleans up IVR navigation and updates the keypad model, but it introduces a couple of correctness risks around DTMF handling and URL parsing. The rest of the changes are reasonably covered by tests.

### Issues

1. **HIGH — `platform/app/services/fsm/instantiation/insti.py` / `_get_key_press_url`**
   - **Problem:** `quote(key, safe="")` percent-encodes DTMF symbols like `#` and `*`, so generated playback URLs will no longer contain the actual key token the downstream prompt builder expects.
   - **Impact:** Pagination prompts for the new `#` / `*` keys can become incorrect or unresolvable at runtime, breaking menu navigation.
   - **Fix:** Keep DTMF keys unescaped in this template replacement path, or explicitly allow `#` and `*` through when building the prompt URL. If encoding is needed elsewhere, isolate it to the blob path only.

2. **HIGH — `platform/app/services/sas_service.py` / `get_url_with_sas`**
   - **Problem:** The URL parsing logic now assumes the incoming blob URL is already structured, but only unquotes each path segment after parsing. This changes behavior for URLs where encoded slashes or spaces appear in the path, and it can misparse container/blob names compared to the previous whole-path decode.
   - **Impact:** SAS generation may fail or produce invalid signatures for existing blob URLs that rely on percent-encoding in the full path.
   - **Fix:** Decode and normalize the URL consistently before splitting, or parse the path in a single canonical step and validate the container/blob segments before generating the SAS token. Add a regression test for an encoded blob path.

3. **MEDIUM — `platform/app/services/fsm/instantiation/pure_audio.py` / `generate_state`**
   - **Problem:** The menu option semantics changed from “next instructions to exit” to “pause/resume”, but the transition logic for key `"0"` was not updated in this diff. That creates a mismatch between the displayed option and the actual behavior.
   - **Impact:** Users may hear a pause/resume prompt while the FSM still behaves like the old instruction flow, causing broken or misleading IVR navigation.
   - **Fix:** Update the `"0"` transition/action wiring in `generate_state` to match the new label, or restore the old label if the state behavior has not changed.

4. **LOW — `platform/app/services/fsm/instantiation/insti.py` / `_Option`**
   - **Problem:** The type annotation was narrowed from `int` to `str`, but the surrounding code still relies on mixed numeric/string handling in several places.
   - **Impact:** This weakens the type contract without fully enforcing string-only keys, making the class interface less accurate than the implementation.
   - **Fix:** Make `_Option.key` an explicit union type if both forms are still valid, or normalize all key creation sites to strings and keep the annotation strict.

### References
- [Python `urllib.parse.quote`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.quote)

### Review Summary

| Severity | Open Issues |
|----------|-------------|
| Critical | 0 |
| High     | 2 |
| Medium   | 1 |
| Low      | 1 |

**Score:** 88/100

**Review by a4i-architect — COMMENTED** [review]
### High-level summary
The navigation-key refactor is directionally good, but there are a few correctness regressions around key handling and one potential parsing bug in SAS URL generation. The main risk is inconsistent handling of numeric vs string DTMF keys now that the IVR system includes `#`/`*` and menu options are normalized to strings.

### Issues

#### 1) **High** — `platform/app/services/fsm/instantiation/insti.py` / `_add_nav_action`
`nav_key` is now typed as `str`, but the helper still builds transition/menu behavior from values that may be consumed elsewhere as numeric identifiers. Because this function is also used for non-pagination navigation keys, coercing everything to strings can break lookups if downstream code expects integers or serializes options differently.

**Fix:** keep the public input flexible (`int | str`) and normalize only at the point of DTMF URL generation; preserve the original key type in the menu mapping or explicitly update all consumers to use string keys end-to-end.

#### 2) **High** — `platform/app/services/fsm/instantiation/insti.py` / `_get_stream_actions`
The page-size change to 8 means the first page can now consume all single-digit DTMF keys, but the current action ordering still appends navigation keys after content keys without guarding collisions. If a page ever contains 8 items plus another navigation action, the UI/menu can expose duplicate or unreachable key bindings.

**Fix:** enforce a per-page key budget before appending nav actions, and make the pagination key insertion conditional on remaining capacity so no key is reused on the same page.

#### 3) **Medium** — `platform/app/services/fsm/instantiation/ivr_constants.py` / module constants
Changing pagination from numeric keys to `#` / `*` is fine, but there is no centralized validation that these keys are supported by all call paths consuming `InputAction(type_=["dtmf"])`. That makes regressions likely if any IVR client or gateway strips or remaps non-digit DTMF symbols.

**Fix:** add a single compatibility check in the FSM instantiation path that validates supported DTMF symbols before generating menu transitions, and fail fast with a logged error if the deployment environment cannot accept `#` / `*`.

#### 4) **Medium** — `platform/app/services/fsm/instantiation/pure_audio.py` / `generate_state`
The repeat-key removal leaves the playback state with only `0` and `9`, but the state description and transition model still imply a menu-driven control surface. This creates a user-facing behavior gap if the audio state was previously relying on repeat as a recovery action.

**Fix:** update the playback-state prompt/description and transition semantics together so the user hears the new available actions instead of the removed repeat behavior.

#### 5) **High** — `platform/app/services/sas_service.py` / `get_url_with_sas`
Parsing the URL before unquoting the full string changes behavior for already-encoded blob paths and can mis-handle URLs containing encoded slashes or special characters. This can generate SAS tokens for the wrong blob path or fail on valid URLs.

**Fix:** parse the URL first, but keep path normalization consistent with the original encoding model; only unquote the path segments that are actually blob name components, and add a regression test for encoded container/blob names.

#### 6) **Low** — `ContentWebApp/src/components/ViewIVR.js` / `createFlowElements`
The display logic now special-cases `"0"` by stringifying only for comparison, but it still renders the original key value for other entries. That means the UI can show inconsistent key formatting when keys are mixed numbers/strings.

**Fix:** normalize the displayed key once in a local variable and render that consistently for all options.

### References
- [RFC 3986 - Uniform Resource Identifier (URI): Generic Syntax](https://www.rfc-editor.org/rfc/rfc3986)
- [Azure Blob Storage SAS overview](https://learn.microsoft.com/en-us/azure/storage/common/storage-sas-overview)

### Review Summary

| Severity | Open Issues |
|----------|-------------|
| Critical | 0 |
| High | 3 |
| Medium | 2 |
| Low | 1 |

**Score:** 88/100

File: platform/app/services/fsm/instantiation/ivr_constants.py (hunk 1 above)
Comment (farmanahmed888) [reply platform/app/services/fsm/instantiation/ivr_constants.py]: done

**Review by N1kunj1998 — APPROVED** [review]
LGTM

**Comment by Soumabha** [comment]
| | Before | After |
| -- | -- | -- |
| Page size | 4 items per page | 8 items per page |
| Content keys | 1–4 | 1–8 |
| Next page | 5 | # |
| Prev page | 7 | * |
| Repeat menu | 8 | removed |
| Back a level | 9 | 9 (unchanged) |

We need to keep repeat menu as it is a user experience enhancement and is especially critical for when a person misses what the IVR just talked about. Pls include the change to keep 'Repeat Menu" preferably in 8 key press like before.

**Review by a4i-architect — COMMENTED** [review]
### High-level summary
The PR mostly modernizes IVR navigation, but it introduces a few correctness regressions around keypad semantics and SAS URL handling. The biggest risk is that the navigation constants and the generated menu behavior are now inconsistent, which can break DTMF routing in production.

### Issues

1. **`platform/app/services/fsm/instantiation/ivr_constants.py` — `number_of_categories_listed_in_one_state` / `next_n_categories_key` / `previous_n_categories_key`**
   - The updated values are internally inconsistent with the existing menu-generation logic and the new tests: the page-size constant is changed to 7, but the tests and PR description expect 8; also the pager keys are now `#`/`*` while the rest of the flow still treats `9` as the back-level key. This will cause pagination and keypad selection to diverge at runtime.
   - **Fix:** Make the constant values and the generator logic agree on a single paging model, then update any dependent menu construction and tests to use that same model end-to-end.

2. **`platform/app/services/fsm/instantiation/insti.py` — `_get_stream_actions`**
   - `key_to_value_mapping` is now keyed by strings, but the function still builds menu items from mixed navigation/content semantics. Because `#` and `*` are introduced as navigation keys, any downstream code expecting integer keys will no longer resolve selected options correctly.
   - **Fix:** Normalize the entire IVR option model around one key type at the boundary (prefer `str` for DTMF), and update all consumers of `menu.options` / `key_to_value_mapping` to treat keys as strings consistently.

3. **`platform/app/services/fsm/instantiation/pure_audio.py` — `generate_state`**
   - Removing the repeat transition also removed the user’s only in-state recovery action for playback. For audio playback, that is a functional regression: if the caller misses part of the menu/audio, there is no longer a way to replay it from the playback state.
   - **Fix:** Replace the removed repeat behavior with another recovery path explicitly supported by the new UX, or keep a replay action on the playback state so callers can re-hear the message without restarting the whole flow.

4. **`platform/app/services/sas_service.py` — `get_url_with_sas`**
   - The URL parsing change now parses the raw URL before decoding the path segments. For already encoded blob names, this can alter path interpretation and generate SAS tokens for the wrong blob path.
   - **Fix:** Decode only the path segments you need after parsing, but preserve the original URL structure for container/blob extraction; add a test covering an encoded blob path to verify the generated SAS URL targets the correct object.

5. **`platform/tests/unit/test_ivr_constants.py` — `test_page_size_is_eight`**
   - The test asserts a page size that contradicts the changed constant in the same PR, so it will fail immediately and block CI.
   - **Fix:** Align the asserted value with the actual constant or correct the constant if 8 is the intended page size.

### References
- [Azure Blob Storage SAS overview](https://learn.microsoft.com/en-us/azure/storage/common/storage-sas-overview)
- [Python `urllib.parse` documentation](https://docs.python.org/3/library/urllib.parse.html)

### Review Summary

| Severity | Open Issues |
|----------|-------------|
| Critical | 0   |
| High | 5   |
| Medium | 0   |
| Low | 0   |

**Score:** 85/100

**Review by a4i-architect — COMMENTED** [review]
### Key Findings

- **High risk regression in IVR navigation constants**: the pagination window changed, but the menu generation logic in the modified files does not clearly show corresponding bound/transition updates. This can break paging behavior or expose the wrong options at boundary states.
- **Potential URL parsing regression in SAS generation**: changing when `unquote` is applied can alter how already-encoded blob paths are parsed, which risks generating invalid SAS URLs for objects with encoded path segments.
- **Type changes are inconsistent across menu option handling**: switching `_Option.key` and nav key mappings from integers to strings is fine only if every caller is updated; otherwise lookups and tests can silently diverge.
- **One test appears to contradict the PR intent**: the new test suite still asserts the repeat option exists in `insti.py`, while the PR description says repeat was removed from the IVR navigation system. That looks like an unresolved expectation mismatch.

### Issues

#### 1) `platform/app/services/fsm/instantiation/ivr_constants.py` — module constants
**Severity: High**

The new pagination constants change the page size and navigation keys, but the PR also leaves `repeat_current_categories_key` defined. That makes the public contract inconsistent with the described removal of repeat navigation and increases the chance of stale callers still relying on it.

**Concrete fix:** remove the constant entirely and update any remaining callers/tests to use the new `#/*` pagination flow only. If the repeat key must remain for compatibility, keep it behind a compatibility alias and document it clearly.

#### 2) `platform/app/services/fsm/instantiation/insti.py` — `_get_stream_actions`
**Severity: High**

The key-mapping changes switch the lookup table from integer keys to string keys, but this function still builds mixed navigation/content semantics. If any downstream consumer still expects numeric keys, menu selection will fail at runtime for existing flows.

**Concrete fix:** normalize the key type at the model boundary, and update the menu/transition consumers to read only one canonical type. Add a compatibility layer only if you must support old serialized FSMs.

#### 3) `platform/app/services/fsm/instantiation/pure_audio.py` — `generate_state`
**Severity: High**

Removing the repeat transition without ensuring the UI/menu text and event handling are aligned can leave the audio playback state with no equivalent way to re-play the current prompt. That is a functional regression if users depended on it for accessibility.

**Concrete fix:** either add the new intended replacement action in this state or update the prompt and input handling so the removed action is not advertised anywhere in the playback flow.

#### 4) `platform/app/services/sas_service.py` — `get_url_with_sas`
**Severity: Medium**

Parsing the raw URL first and only then unquoting each path segment changes behavior for encoded blob names. This can fail for blobs whose path contains reserved characters or already-encoded segments, producing invalid container/blob extraction.

**Concrete fix:** keep path parsing and decoding consistent for the exact blob URL format you expect, and add tests for encoded blob names and nested paths.

#### 5) `platform/tests/unit/test_insti_nav_keys.py` — `TestRepeatKeyKept.test_repeat_option_in_menu_on_key_eight`
**Severity: High**

This test still asserts the repeat option exists even though the PR description says repeat navigation was removed. That means the test suite is codifying the old behavior and will mask a real mismatch if the implementation removes it elsewhere.

**Concrete fix:** update the test to assert the new intended behavior for key `8`, or remove the test if repeat is no longer supported.

#### 6) `platform/tests/unit/test_ivr_constants.py` — `test_repeat_key_constant_is_eight`
**Severity: Medium**

This test preserves the repeat-key contract while the PR narrative says the constant should be removed. It creates a false sense of coverage around a deprecated behavior.

**Concrete fix:** delete the test if the constant is removed, or rename/reframe it as a backward-compatibility test if the constant must remain temporarily.

### References
- [Python `urllib.parse` documentation](https://docs.python.org/3/library/urllib.parse.html)

### Review Summary

| Severity | Open Issues |
|----------|-------------|
| Critical | 0   |
| High | 4   |
| Medium | 2   |
| Low | 0   |

**Score:** 86/100

File: platform/app/services/fsm/instantiation/pure_audio.py [hunk 2]
```diff
@@ -171,12 +169,4 @@ def generate_state(
                 actions=[],
             )
         )
-        fsm.add_transition(
```
Comment (Soumabha) [inline platform/app/services/fsm/instantiation/pure_audio.py]: Needs to be added back to allow repetition of the menu

File: platform/app/services/fsm/instantiation/pure_audio.py [hunk 3]
```diff
@@ -126,9 +125,8 @@ def generate_state(
         actions.append(InputAction(type_=["dtmf"], eventApi="/input", timeOut=10))
 
         options = [
-            _Option(key=8, value="repeat"),
-            _Option(key=9, value="exit"),
-            _Option(key=0, value="next (instructions to exit)"),
+            _Option(key="9", value="exit"),
```
Comment (Soumabha) [inline platform/app/services/fsm/instantiation/pure_audio.py]: We need to add back 8 for the repeat option.

**Review by Soumabha — CHANGES_REQUESTED** [review]
Reviewed the latest additions for restoration of 8. A couple of pending pointers.

**Review by a4i-architect — COMMENTED** [review]
### Key findings

- **High:** The pagination redesign appears internally inconsistent: the constants file now exposes `previous_n_categories_key`, but the diff in `ivr_constants.py` still only shows `previous_n_categories_key` being introduced indirectly via tests, and the runtime code in `insti.py` relies on string keys everywhere. This needs a careful end-to-end check so menu generation and transition lookup stay aligned.
- **High:** The removal of the repeat-menu behavior in `pure_audio.py` is only partially reflected in the tests; the new audio-state menu still labels key `8` as repeat, which can conflict with the stated removal of repeat navigation.
- **Medium:** `sas_service.py` changed URL parsing behavior; decoding path segments individually is safer for encoded blob names, but it can break URLs with encoded container/blob separators if not handled consistently.
- **Low:** `ViewIVR.js` now special-cases `"0"` instead of numeric `0`, which is fine for the new string-key model, but it should be kept consistent across all menu renderers.

### Issues

#### 1) Pagination key handling may diverge between constants, menu generation, and tests
**File:** `platform/app/services/fsm/instantiation/ivr_constants.py`  
**Function:** module-level constants used by menu builders

- The PR changes navigation semantics to `#` / `*` and expands the page size to 7, but the runtime diff shown here does not include a corresponding update in the menu-generation code path that consumes these constants. That creates a risk that the UI and FSM disagree about which keys are actionable on each page.
- **Concrete fix:** verify every consumer of `number_of_categories_listed_in_one_state`, `next_n_categories_key`, and `previous_n_categories_key` is updated together, and add a single integration test that asserts the generated menu and transition map use the same symbols on first, middle, and last pages.

**Severity:** HIGH

#### 2) Audio playback still exposes repeat semantics on key 8
**File:** `platform/app/services/fsm/instantiation/pure_audio.py`  
**Function:** `generate_state`

- The PR description says repeat is removed, but the modified audio state still binds key `8` to a repeat action in the menu. That means the user-facing behavior and the stated design change do not match.
- **Concrete fix:** either remove the `8` option and any self-loop transition, or rename/rework it so the menu label and transition behavior match the new navigation model. Update the audio navigation test accordingly.

**Severity:** HIGH

#### 3) SAS URL parsing can mis-handle already-encoded blob paths
**File:** `platform/app/services/sas_service.py`  
**Function:** `get_url_with_sas`

- Switching from decoding the full URL to decoding path segments individually changes how special characters are interpreted. This can produce incorrect container/blob extraction for some valid Azure blob URLs, especially when the original path contains encoded separators or reserved characters.
- **Concrete fix:** normalize the URL once, then parse container/blob components using Azure-safe path handling; add tests for encoded container/blob names and URLs that already include SAS parameters.

**Severity:** MEDIUM

#### 4) View rendering now depends on stringified key `0` only
**File:** `ContentWebApp/src/components/ViewIVR.js`  
**Function:** `createFlowElements`

- The new comparison handles string keys, but it introduces a hard dependency on one exact representation of zero. If any caller still emits numeric `0`, the display behavior will diverge.
- **Concrete fix:** standardize key serialization at the data boundary and keep the renderer consistent with that contract; add a small UI test for both rendered zero and non-zero keys.

**Severity:** LOW

### Review Summary

| Severity | Open Issues |
|----------|-------------|
| Critical | 0 |
| High | 2 |
| Medium | 1 |
| Low | 1 |

**Score:** 89/100

**Review by a4i-architect — COMMENTED** [review]
### Key findings

- The navigation key refactor introduces a couple of **behavioral mismatches** between constants, generated menus, and tests.
- There is also a **possible broken route in SAS URL parsing** if encoded paths are passed through unchanged.
- The UI change in `ViewIVR.js` is safe, but the backend changes need tighter consistency to avoid menu corruption or failing lookups.

### Issues

#### 1) **High** — `platform/app/services/fsm/instantiation/ivr_constants.py` / module constants
The pagination constants now use `#` and `*`, but `repeat_current_categories_key` is still present and the tests indicate it is still expected. This makes the PR internally inconsistent with the stated removal of repeat handling and risks leaving dead navigation behavior in place.

**Fix:** Remove the repeat constant entirely if the feature is being deleted, and update any callers/tests to stop referencing it. If repeat is still required, keep the constant but restore the associated state transitions so the system stays coherent.

#### 2) **High** — `platform/app/services/fsm/instantiation/insti.py` / `_add_nav_action`
Changing `key_to_value_mapping` to store only string keys can break any downstream code that still looks up options with integers. This is a functional compatibility risk because the rest of the FSM/test surface previously used numeric keys in several places.

**Fix:** Normalize keys at the boundary in one place only, and keep the public menu model consistent. If string keys are the new canonical form, update all consumers to use strings and add a compatibility conversion layer where lookups are performed.

#### 3) **Medium** — `platform/app/services/sas_service.py` / `get_url_with_sas`
Parsing the URL before unquoting the full path changes how encoded blob names are interpreted. If the original URL contains encoded slashes or special characters, splitting the path first can produce the wrong container/blob name and fail SAS generation.

**Fix:** Preserve the existing decode-then-parse behavior for the blob path, or decode each path segment only after confirming the container/blob boundary is correct. Add a unit test covering an encoded blob name.

#### 4) **Medium** — `platform/tests/unit/test_ivr_constants.py` / module-level tests
The new constants tests hard-code the repeat key as still existing even though the PR description says it was removed. That means the test suite is validating the old contract, not the new one.

**Fix:** Align the tests with the intended contract: either assert the constant is gone and no code imports it, or keep the feature and update the PR description to match the implementation.

#### 5) **Low** — `ContentWebApp/src/components/ViewIVR.js` / `createFlowElements`
The rendering change only special-cases `"0"` as a string, which is fine for the current backend refactor, but it makes the UI depend on a backend-specific key normalization choice.

**Fix:** Normalize option keys once at the data boundary in the frontend model layer so the presentation component doesn’t need to know whether keys arrive as numbers or strings.

### References
- [urllib.parse.urlparse — Python docs](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse)

### Review Summary

| Severity | Open Issues |
|----------|-------------|
| Critical | 0   |
| High | 2   |
| Medium | 2   |
| Low | 1   |

**Score:** 88/100

**Review by Soumabha — APPROVED** [review]
